src/ui/app_header/RadialCircularMenu.hpp

```diff
@@ -0,0 +1,283 @@
+        for (auto i = firstButtonIndex; i < _buttons.size(); ++i) {
+            const auto &button = _buttons[i];
+            if ((cumulativeLength + button.size + _padding) > arcLength || (button.isNewRow && i > firstButtonIndex)) {
+                break;
+            }
+            ++buttonCount;
+            maxButtonSize = std::max(button.size, maxButtonSize);
+            cumulativeLength += button.size + _padding;
+        }
+        return { buttonCount, maxButtonSize };
+    }
+
+public:
+    explicit RadialCircularMenu(ImVec2 menuSize = { 100, 100 }, float startAngle = 0.f, float stopAngle = 360.f, float extraRadius = 0.f, float animationSpeed = .25f, float timeOut = 0.5f)
+        : _menuSize(menuSize), _startAngle(startAngle), _stopAngle(stopAngle), _extraRadius(extraRadius), _animationSpeed(animationSpeed), _timeOut(timeOut) {
+        updateAndDraw();
+    }
+
+    template<bool transparent = false, bool newRow = false>
+    void addButton(std::string label, auto onClick, float buttonSize = -1.f, std::string toolTip = "") {
+        if (_animationProgress > 0.f) { // we do not allow to add buttons when the popup is already open or animating
+            return;
+        }
+        _buttons.emplace_back(label, buttonSize, std::move(onClick), nullptr, std::move(toolTip), transparent, newRow);
+        _isOpen = true;
+    }
+
+    template<bool transparent = false, bool newRow = false>
+    void addButton(std::string label, auto onClick, ImFont *font, std::string toolTip = "") {
+        if (_animationProgress > 0.f) { // we do not allow to add buttons when the popup is already open or animating
+            return;
+        }
+        float buttonSize = 0.f;
+        if (font == nullptr) {
+            buttonSize = ImGui::CalcTextSize(label.c_str()).y + 2.f * _padding;
+        } else {
+            buttonSize = font->FontSize + 2.f * _padding;
+        }
+        _buttons.emplace_back(label, buttonSize, std::move(onClick), font, std::move(toolTip), transparent, newRow);
+        _isOpen = true;
+    }
+
+    [[nodiscard]] bool isOpen() const noexcept {
+        return _isOpen;
+    }
+
+    void forceClose() {
+        _isOpen            = false;
+        _animationProgress = 0.0;
+        _buttons.clear();
+        ImGui::CloseCurrentPopup();
+    }
+
+    void updateAndDraw() {
+        static float  timeOutOfRadius = 0.0f;
+        static ImVec2 centre{ -1.f, -1.f };
+
+        const float   deltaTime = ImGui::GetIO().DeltaTime;
+        _animationProgress      = _isOpen ? std::min(1.0f, _animationProgress + deltaTime / _animationSpeed) : std::max(0.0f, _animationProgress - deltaTime / _animationSpeed);
+
+        if (!_isOpen && _animationProgress <= 0.f) {
+            centre = { -1.f, -1.f };
+            if (!_buttons.empty()) {
+                _buttons.clear();
+            }
+            return;
+        } else if (_isOpen && (centre.x < 0.f || centre.y < 0.f)) {
+            centre = ImGui::GetMousePos();
```

Comment:
It's probably better to have a fixed well positioned center, otherwise when entering the window from the top edge, the center will be very close to it and the top icon will be partially outside of the window.

Reply:
Will tie it to the calling button-centre.